src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java

```diff
@@ -183,13 +188,166 @@ private void commitEntriesToStorage(Map<WALEntry, WALEntry> entries) {
     }
   }
 
+  /**
+   * Extract the latest offset and file path from the WAL file.
+   * Attempt with the most recent WAL file, and fall back to the old file if it's not applicable.
+   *
+   * <p> The old WAL is used when the most recent WAL file has already been truncated
+   * and is not existent, which may happen when the connector has flushed all records,
+   * and is restarted. </p>
+   *
+   * @return the latest offset and filepath from the WAL file or null
+   */
+  @Override
+  public FilePathOffset extractLatestOffset() {
+    String oldWALFile = logFile + TRUNCATED_LOG_EXTENSION;
+    try {
+      FilePathOffset latestOffset = null;
+      if (storage.exists(logFile)) {
+        log.trace("Restoring offset from WAL file: {}", logFile);
+        if (reader == null) {
+          reader = newWalFileReader();
+        } else {
+          // reset read position after apply()
+          reader.seekToFirstRecord();
+        }
+        List<String> committedFileBatch = getLastFilledBlockFromWAL(reader);
+        // At this point the committedFilenames list will contain the
+        // filenames in the last BEGIN-END block of the file. Find the latest offsets among these.
+        latestOffset = getLatestOffsetFromList(committedFileBatch);
+      }
+
+      // attempt to use old log file if recent WAL is empty or non-existent
+      if (latestOffset == null && storage.exists(oldWALFile)) {
+        log.trace("Could not find offset in log file {}. Using {} instead", logFile, oldWALFile);
+        try (Reader oldFileReader =
+            new WALFile.Reader(conf.getHadoopConfiguration(), Reader.file(new Path(oldWALFile)))
+        ) {
+          List<String> committedFileBatch = getLastFilledBlockFromWAL(oldFileReader);
+          latestOffset = getLatestOffsetFromList(committedFileBatch);
+        }
+      }
```

Comment:
this looks like some repeated code that we can factor out for readability i.e

`private FilePathOffset extractLatestOffset(String logFile);`

WDYT?

Reply:
Close, but the difference in initializing the old vs new readers needs to be kept, so not sure if it would be worth it for the two other lines (226, 227). For the recent WAL we use the logic above: 
```
        if (reader == null) {
          reader = newWalFileReader();
        } else {
          // reset read position after apply()
          reader.seekToFirstRecord();
        }
```
For the old WAL we always need the new instance. Additionally the try-with-resource block (223,224) is only applicable for the old file since it closes the reader after the block, for the recent WAL we need to keep it open.